underscore.js

```diff
@@ -277,56 +277,33 @@
     return _.find(obj, _.matcher(attrs));
   };
 
-  // Return the maximum element (or element-based computation).
-  _.max = function(obj, iteratee, context) {
-    var result = -Infinity, lastComputed = -Infinity,
-        value, computed;
-    if (iteratee == null && obj != null) {
-      obj = isArrayLike(obj) ? obj : _.values(obj);
-      for (var i = 0, length = obj.length; i < length; i++) {
-        value = obj[i];
-        if (value > result) {
-          result = value;
-        }
-      }
-    } else {
-      iteratee = cb(iteratee, context);
-      _.each(obj, function(value, index, list) {
-        computed = iteratee(value, index, list);
-        if (computed > lastComputed || computed === -Infinity && result === -Infinity) {
+  // Generator function to create the max and min functions
+  var createExtremumFinder = function(max) {
+    return function(obj, iteratee, context) {
+      var keys = !isArrayLike(obj) && _.keys(obj),
+          length = (keys || obj).length,
+          result = max ? -Infinity : Infinity,
+          lastComputed = result,
+          found = false;
+      iteratee = iteratee == null ? null : cb(iteratee, context);
```

Comment:
When I was playing around with it a while ago I found the `identity` iteratee was often faster than all the branching